### ISSUE TITLE
[v3] client_credentials should not require a user

### ISSUE THREAD

**Issue by fkammer** [oauthjs/node-oauth2-server#356]
As @romainfrancez already pointed out in #202 for v2, RFC states for grant type `client_credentials`:

> [...] Client credentials are used as an authorization grant
> typically when the client is acting on its own behalf [...]
> -- https://tools.ietf.org/html/rfc6749#section-1.3.4

Right now it's required to provide the function `getUserFromClient`. I don't think it's wrong to have this function in general as there are use-cases where you want to authenticate a user just by client credentials. On the other hand there are use-cases where the client really isn't acting on behalf of an user but as the client itself.
I'd like to change the flow of the `client_credentials` grant type to allow the creation of an access token in the context of an user as well as in the context of a client.

I suggest the following changes and would be more than happy to create a PR:
 - Add optional parameter `context` to all functions inside `saveToken` (e.g. `generateAccessToken`, `validateScope` etc.) which could be `'user'` (default) or `'client'` to allow different access token types / storage for both contexts without introducing new functions.
 - Make implementation of `getUserFromClient` optional. If the function isn't implemented or doesn't return a user, `context` is set to `'client'`.
- Update the authenticate handler allowing to pass the expected context for the endpoint. Currently `AuthenticateHandler.prototype.getAccessToken` expects that an user object is returned by `this.model.getAccessToken`. This should be dependent on the value of `context`:
   - `context == 'user'` (default): `user` object expected;
   - `context == 'client'`: `client` object expected;
   - `context == 'user client'`: `user` or `client` object expected.

Should be possible to implement all this in a backwards-compatible way.

**Comment by fkammer** [comment]
@thomseddon @mjsalinger any thoughts on this?

**Comment by maxtruxa** [comment]
I think `getUserFromClient` being required is actually correct.
i.e. the client authorizes itself to access the data its user is allowed to access.

If you don't need `user` just return an empty object from `getUserFromClient`:
```
function getUserFromClient() {
    return {};
}
```

**Comment by maxtruxa** [comment]
I think `getUserFromClient` being required is actually correct.

Looking at one of the "typical" grants, a client doesn't actually own any resources, but it can request access the resources owned by a resource owner.

Compare that to the client credentials grant:
As you correctly quoted:
> [...] the client is acting on its own behalf [...]
But even if the client can authorize itself... without being a resource owner (aka "user") itself, there are no resources for the client to access.

So, basically the design approach behind `getUserFromClient` assumes that your client *is a resource owner itself*.

By the way, if you don't need `user` just return an empty object from `getUserFromClient`:
```
function getUserFromClient() {
    return {};
}
```

**Comment by bencmbrook** [comment]
> Add optional parameter context to all functions inside saveToken (e.g. generateAccessToken, validateScope etc.) which could be 'user' (default) or 'client' to allow different access token types / storage for both contexts without introducing new functions.

This would be an awesome feature. I'm currently trying to implement custom code to get this context. Any suggestions for getting the `grant_type` of the request inside my `saveToken` function?
